internal/engine/watchmanager.go

```diff
@@ -0,0 +1,154 @@
+package engine
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/windmilleng/tilt/internal/ignore"
+	"github.com/windmilleng/tilt/internal/logger"
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/watch"
+)
+
+type manifestFilesChangedAction struct {
+	manifestName model.ManifestName
+	files        []string
+}
+
+func (manifestFilesChangedAction) Action() {}
+
+type manifestNotifyCancel struct {
+	manifest model.Manifest
+	notify   watch.Notify
+	cancel   func()
+}
+
+type WatchManager struct {
+	watches        map[model.ManifestName]manifestNotifyCancel
+	fsWatcherMaker FsWatcherMaker
+	// TODO(dmiller) do I need a mutex here?
+}
+
+func NewWatchManager(watcherMaker FsWatcherMaker) *WatchManager {
+	return &WatchManager{
+		watches:        make(map[model.ManifestName]manifestNotifyCancel),
+		fsWatcherMaker: watcherMaker,
+	}
+}
+
+func (w *WatchManager) diff(ctx context.Context, state store.EngineState) (setup []model.Manifest, teardown []model.Manifest) {
+	setup = []model.Manifest{}
+	teardown = []model.Manifest{}
+	manifestsToProcess := make(map[model.ManifestName]model.Manifest, len(state.ManifestStates))
+	for i, m := range state.ManifestStates {
+		manifestsToProcess[i] = m.Manifest
+	}
+	for n, state := range state.ManifestStates {
+		_, ok := w.watches[n]
+		if !ok {
+			setup = append(setup, state.Manifest)
+		}
+		delete(manifestsToProcess, n)
+	}
+
+	for _, m := range manifestsToProcess {
+		teardown = append(teardown, m)
+	}
+
+	return setup, teardown
+}
+
+func (w *WatchManager) OnChange(ctx context.Context, st *store.Store) {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	setup, teardown := w.diff(ctx, state)
+
+	for _, m := range teardown {
```

Comment:
maybe teardown should be an array of watches, rather than an array of Manifests? it seems weird that you have to look them up again.

Reply:
True but we need the manifest name to subsequently delete the manifestWatchCancel pair from the map.